### ISSUE TITLE
Cause build error on s390x

### ISSUE THREAD

**Issue by kiszk** [pytorch/pytorch#112017]
### 🐛 Describe the bug

```
% python setup.py install
...
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:405:60: error: ‘struct at::vec::ZVECTOR::Vectorized<c10::complex<float> >’ has no member named ‘atanh’; did you mean ‘atan’?
  405 |         [=](Vectorized<scalar_t> self_vec){return self_vec.atanh();});
...
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/detail/FunctionTraits.h:9:8: error: invalid use of incomplete type ‘struct function_traits<auto (at::native::ZVECTOR::atanh_kernel(at::TensorIteratorBase&)::<lambda()>::<lambda()>::<lambda(at::vec::ZVECTOR::Vectorized<c10::complex<float> >)>::*)(at::vec::ZVECTOR::Vectorized<c10::complex<float> >) const>’
    9 | struct function_traits : public function_traits<decltype(&T::operator())> {
      |        ^~~~~~~~~~~~~~~
...
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/native/cpu/Loops.h:203:36: error: ‘arity’ is not a member of ‘traits’ {aka ‘function_traits<at::native::ZVECTOR::atanh_kernel(at::TensorIteratorBase&)::<lambda()>::<lambda()>::<lambda(at::vec::ZVECTOR::Vectorized<c10::complex<float> >)>&>’}
  203 |   constexpr int ntensors = traits::arity + 1;
      |                                    ^~~~~
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/native/cpu/Loops.h:205:27: error: size of array is not an integral constant-expression
  205 |   char* C10_RESTRICT data[ntensors];
      |                           ^~~~~~~~
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/native/cpu/Loops.h:213:41: error: no matching function for call to ‘dereference_vec<traits>(char* __restrict__*, Vec&, int64_t&, int64_t&)’
  213 |     auto args1 = dereference_vec<traits>(&data[1], opt_scalar, S, i);
      |                  ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
/home1/ishizaki/PyTorch/main-lastest/aten/src/ATen/native/cpu/Loops.h:78:1: error: no type named ‘ArgsTuple’ in ‘struct function_traits<at::native::ZVECTOR::atanh_kernel(at::TensorIteratorBase&)::<lambda()>::<lambda()>::<lambda(at::vec::ZVECTOR::Vectorized<c10::complex<float> >)>&>’
...
```

### Versions
commit: 88244cd7a989d444d7bd4fe9cc9e336df0fb05ef

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 23.04 (s390x)
GCC version: (Ubuntu 12.3.0-1ubuntu1~23.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: glibc-2.37

Python version: 3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-6.4.6-s390x-with-glibc2.37
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                    s390x
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Big Endian
CPU(s):                          96
On-line CPU(s) list:             0-95
Vendor ID:                       IBM/S390
Model name:                      -
Machine type:                    3931
Thread(s) per core:              1
Core(s) per socket:              8
Socket(s) per book:              2
Book(s) per drawer:              4
Drawer(s):                       4
CPU dynamic MHz:                 5200
CPU static MHz:                  5200
BogoMIPS:                        26315.00
Dispatching mode:                horizontal
Flags:                           esan3 zarch stfle msa ldisp eimm dfp edat etf3eh highgprs te vx vxd vxe gs vxe2 vxp sort dflt vxp2 nnpa pcimio sie
Hypervisor:                      PR/SM
Hypervisor vendor:               IBM
Virtualization type:             full
L1d cache:                       12 MiB (96 instances)
L1i cache:                       12 MiB (96 instances)
L2 cache:                        32 MiB (1 instance)
L3 cache:                        256 MiB
NUMA node(s):                    1
NUMA node0 CPU(s):               0-95
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Not affected
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; etokens
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.1
[pip3] optree==0.9.2
[conda] Could not collect
```
cc @malfet @seemethere

**Comment by mikaylagawarecki** [comment]
Could you list your environment with the collect_env script please

**Comment by kiszk** [comment]
Sorry for missing it. Just attached.

**Comment by kiszk** [comment]
I just realized that one of failure was fixed by https://github.com/pytorch/pytorch/pull/111653

**Comment by kiszk** [comment]
Fixed by #111653
